perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -266,57 +346,74 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
     object_name = '%s_%d' % (object_prefix, i)
 
-    # Ready to go!
-    start_time = time.time()
-
-    # We write the object to storage provider and if successful we place the
-    # name of the object to the objects_written list. When a write fails, the
-    # object won't be placed into the objects_written list, and then it's up to
-    # the caller to decide if they want to tolerate the failure and accept
-    # the results, depending on test scenarios.
     try:
-      if _useBotoApi(storage_schema):
-        object_path = '%s/%s' % (bucket, object_name)
-        object_uri = boto.storage_uri(object_path, storage_schema)
-        if host_to_connect is not None:
-          object_uri.connect(host=host_to_connect)
-
-        object_uri.set_contents_from_string(payload_string)
-      else:
-        _AZURE_BLOB_SERVICE.put_block_blob_from_bytes(bucket, object_name,
-                                                      bytes(payload_bytes))
-
-      latency = time.time() - start_time
+      _, latency = WriteObjectFromBuffer(
+          storage_schema, bucket, object_prefix,
+          payload_bytes, payload_string, size,
+          host_to_connect=host_to_connect)
 
+      objects_written.append(object_name)
       if latency_results is not None:
         latency_results.append(latency)
-
       if bandwidth_results is not None and latency > 0.0:
         bandwidth_results.append(size / latency)
+    except Exception as e:
+      logging.info('Caught exception %s while writing object %s' %
+                   (e, object_name))
 
-      objects_written.append(object_name)
-    except:
-      logging.exception('Caught exception while writing object %s.',
-                        object_name)
+
+def WriteObjectFromBuffer(storage_schema, bucket, object,
+                          payload_bytes, payload_string, size,
+                          host_to_connect=None):
+  """Write an object to a storage provider.
+
+  Exceptions are propagated to the caller, which can decide whether to
+  tolerate them or not.
+
+  Args:
+    storage_schema: The address schema identifying a storage. e.g., "gs"
+    bucket: Name of the bucket to write to.
+    object: Name of the object.
+    payload_bytes: A bytearray to transfer.
+    payload_string: A string version of payload_bytes.
+    size: The size of the objects to write, in bytes. If size is
+      less than the size of payload_{bytes,string}, then this will
+      transfer the prefix of payload_{bytes,string} of the desired size.
+    host_to_connect: An optional endpoint string to connect to.
+
+  Returns:
+    a tuple (start_time, latency).
+
+  """
+
+  # Ready to go!
+  start_time = time.time()
+
+  if _useBotoApi(storage_schema):
+    object_path = '%s/%s' % (bucket, object)
+    object_uri = boto.storage_uri(object_path, storage_schema)
+    if host_to_connect is not None:
+      object_uri.connect(host=host_to_connect)
+
+    object_uri.set_contents_from_string(payload_string[:size])
```

Comment:
Python does "slice-by-copy" so payload_string[:size] would generate a new array object and has a copy of the original array, which is very inefficient here. I think your design calls for sharing the read-only global buffer.

We need a "view" into the string or byte array, perhaps the memoryview class fits well here, please take a look at that.